File: leetcode_java/89. Gray Code.java
class Solution {
    public List<Integer> grayCode(int n) {
        List<Integer> res = new ArrayList<Integer>() {{ add(0); }};
        for (int i = 0; i < n; i++) {
            int head = 1 << i;
            for (int j = res.size() - 1; j >= 0; j--)
                res.add(head + res.get(j));
        }
        return res;
    }
}

File: leetcode_python/89. Gray Code.py
class Solution:
    def grayCode(self, n: int) -> List[int]:
        res = [0]
        for i in range(n):
            head = 1 << i
            for j in range(len(res) - 1, -1, -1):
                res.append(head + res[j])
        return res